### PR TITLE
Upgrade to use new IJ ActionUpdateThread API

### LINKED ISSUE #610: Event Dispatch Thread errors
A number of users have reported issues with `EDT`. Collected here: 

`Access is allowed from Event Dispatch Thread (EDT) only`
https://github.com/sourcegraph/jetbrains/issues/344
https://github.com/sourcegraph/jetbrains/issues/368
https://github.com/sourcegraph/jetbrains/issues/469
https://github.com/sourcegraph/jetbrains/issues/492
https://github.com/sourcegraph/jetbrains/issues/524
https://github.com/sourcegraph/jetbrains/issues/556
https://github.com/sourcegraph/jetbrains/issues/563
https://github.com/sourcegraph/jetbrains/issues/584
https://github.com/sourcegraph/jetbrains/issues/622
https://github.com/sourcegraph/jetbrains/issues/676

`EventQueue.isDispatchThread()=false`
https://github.com/sourcegraph/jetbrains/issues/465
https://github.com/sourcegraph/jetbrains/issues/467
https://github.com/sourcegraph/jetbrains/issues/572

`ActionUpdateThread.OLD_EDT is deprecated and going to be removed soon`
https://github.com/sourcegraph/jetbrains/issues/441
https://github.com/sourcegraph/jetbrains/issues/498
https://github.com/sourcegraph/jetbrains/issues/499
https://github.com/sourcegraph/jetbrains/issues/539
https://github.com/sourcegraph/jetbrains/issues/575
https://github.com/sourcegraph/jetbrains/issues/641
https://github.com/sourcegraph/jetbrains/issues/660
https://github.com/sourcegraph/jetbrains/issues/670
https://github.com/sourcegraph/jetbrains/issues/672
https://github.com/sourcegraph/jetbrains/issues/673
https://github.com/sourcegraph/jetbrains/issues/767
https://github.com/sourcegraph/jetbrains/issues/768
https://github.com/sourcegraph/jetbrains/issues/769
https://github.com/sourcegraph/jetbrains/issues/782
https://github.com/sourcegraph/jetbrains/issues/810
https://github.com/sourcegraph/jetbrains/issues/824
https://github.com/sourcegraph/jetbrains/issues/825
https://github.com/sourcegraph/jetbrains/issues/826
https://github.com/sourcegraph/jetbrains/issues/827
https://github.com/sourcegraph/jetbrains/issues/828
https://github.com/sourcegraph/jetbrains/issues/829
https://github.com/sourcegraph/jetbrains/issues/830
https://github.com/sourcegraph/jetbrains/issues/831
https://github.com/sourcegraph/jetbrains/issues/832
https://github.com/sourcegraph/jetbrains/issues/833
https://github.com/sourcegraph/jetbrains/issues/861
https://github.com/sourcegraph/jetbrains/issues/869

`xx ms to call on EDT CodyStatusBarActionGroup`
https://github.com/sourcegraph/jetbrains/issues/631
https://github.com/sourcegraph/jetbrains/issues/665
https://github.com/sourcegraph/jetbrains/issues/878
https://github.com/sourcegraph/jetbrains/issues/913

### PR DESCRIPTION
**Pull request by pkukielka**
## Changes

Fixes https://github.com/sourcegraph/jetbrains/issues/610
`ActionUpdateThread.OLD_EDT is deprecated and going to be removed soon`

Why we get that issue?
https://plugins.jetbrains.com/docs/intellij/basic-action-system.html#principal-implementation-overrides
https://github.com/JetBrains/intellij-community/blob/idea/233.14475.28/platform/editor-ui-api/src/com/intellij/openapi/actionSystem/ActionUpdateThread.java

New API was become required in 2024 EAP, and every IntelliJ `Action` must define if `update` should be running on `EDT` or `BGT` thread. Unfortunately that API was not present before `2022.2`, so we need to upgrade a bit.

## Test plan

This change requires full manual QA

### CODE REVIEW

File: src/main/kotlin/com/sourcegraph/common/ui/DumbAwareBGTAction.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+  constructor(text: @NlsActions.ActionText String?) : super(text)
+
+  constructor(
+      text: @NlsActions.ActionText String?,
+      description: @NlsActions.ActionDescription String?,
+      icon: Icon?
+  ) : super(text, description, icon)
+
+  @Suppress("MissingRecentApi")
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/common/ui/DumbAwareBGTAction.kt]: nice 🚀

Comment (mkondratek) [reply src/main/kotlin/com/sourcegraph/common/ui/DumbAwareBGTAction.kt]: btw, can you provide a comment on this class? to explain why we need it, why we need to suppress this warning, and when can we drop it.

File: src/main/kotlin/com/sourcegraph/common/UpgradeToCodyProNotification.kt [hunk 2]
```diff
@@ -23,7 +23,7 @@ private constructor(title: String, content: String, shouldShowUpgradeOption: Boo
   init {
     icon = Icons.CodyLogo
     val learnMoreAction: AnAction =
-        object : DumbAwareAction("Learn more") {
+        object : DumbAwareBGTAction("Learn More") {
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/common/UpgradeToCodyProNotification.kt]: no strong feelings but `Learn more` looks better imho

File: src/main/kotlin/com/sourcegraph/cody/ui/AutoGrowingTextArea.kt [hunk 3]
```diff
@@ -1,15 +1,11 @@
 package com.sourcegraph.cody.ui
 
 import com.intellij.ide.ui.laf.darcula.ui.DarculaTextAreaUI
-import com.intellij.openapi.actionSystem.AnAction
-import com.intellij.openapi.actionSystem.AnActionEvent
-import com.intellij.openapi.actionSystem.CustomShortcutSet
-import com.intellij.openapi.actionSystem.KeyboardShortcut
-import com.intellij.openapi.actionSystem.ShortcutSet
-import com.intellij.openapi.project.DumbAwareAction
+import com.intellij.openapi.actionSystem.*
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/ui/AutoGrowingTextArea.kt]: I assume your IntelliJ settings changed it. I do not like wildcard imports ([StackOverflow](https://stackoverflow.com/a/147461))

File: src/main/kotlin/com/sourcegraph/cody/ui/AutoGrowingTextArea.kt [hunk 4]
```diff
@@ -28,11 +24,10 @@ class AutoGrowingTextArea(private val minRows: Int, maxRows: Int, outerPanel: JC
   val textArea: JBTextArea
   val scrollPane: JBScrollPane
   private val initialPreferredSize: Dimension
-  private val autoGrowUpToRow: Int
-  val undoManager = UndoManager()
+  private val autoGrowUpToRow: Int = maxRows + 1
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/ui/AutoGrowingTextArea.kt]: what changed here?

File: src/main/kotlin/com/sourcegraph/cody/statusbar/ReportCodyBugAction.kt [hunk 5]
```diff
@@ -2,9 +2,9 @@ package com.sourcegraph.cody.statusbar
 
 import com.intellij.ide.BrowserUtil
 import com.intellij.openapi.actionSystem.AnActionEvent
-import com.intellij.openapi.project.DumbAwareAction
+import com.sourcegraph.common.ui.DumbAwareBGTAction
 
-class ReportCodyBugAction : DumbAwareAction("Open GitHub To Report Cody Issue") {
+class ReportCodyBugAction : DumbAwareBGTAction("Open GitHub To Report Cody Issue") {
   override fun actionPerformed(p0: AnActionEvent) {
     BrowserUtil.open("https://github.com/sourcegraph/jetbrains/issues/new?template=bug_report.yml")
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/statusbar/ReportCodyBugAction.kt]: where is it used? this action? don't we need to put the label param btw?

File: src/main/kotlin/com/sourcegraph/cody/initialization/TrialEndedNotification.kt [hunk 6]
```diff
@@ -39,6 +43,9 @@ class TrialEndedNotification(val disposable: Disposable) :
             disposable.dispose()
             notification.expire()
           }
+
+          @Suppress("MissingRecentApi")
+          override fun getActionUpdateThread(): ActionUpdateThread = ActionUpdateThread.BGT
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/initialization/TrialEndedNotification.kt]: IIUC this is the other case (next to `DumbAwareAction`) where we need to define `getActionUpdateThread`, right? 

1. can we have a similar solution - `NotificationBGTAction`? 
2. can you add a CI check that verifies that we DO not use `NotificationAction` or `DumbAwareAction` in any other place than the corresponding abstract classes?

File: gradle.properties [hunk 7]
```diff
@@ -10,12 +10,12 @@ pluginSinceBuild=221.1
...
 # IntelliJ Platform Properties -> https://github.com/JetBrains/gradle-intellij-plugin#intellij-platform-properties
 platformType=IC
-platformVersion=2022.1
+platformVersion=2022.2
 # Plugin Dependencies -> https://plugins.jetbrains.com/docs/intellij/plugin-dependencies.html
 # Example: platformPlugins = com.intellij.java, com.jetbrains.php:203.4449.22
 platformPlugins=Git4Idea,PerforceDirectPlugin
 # Java language level used to compile sources and to generate the files for - Java 11 is required for 2020.3 <= x < 2022.2
-javaVersion=11
+javaVersion=17
```
Comment (mkondratek) [inline gradle.properties]: 🕵️

**Review by mkondratek — COMMENTED** [review]
comments 🗯️

File: src/main/kotlin/com/sourcegraph/common/UpgradeToCodyProNotification.kt (hunk 2 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/common/UpgradeToCodyProNotification.kt]: That is against IntelliJ action names guidelines, it's highlighted in the IDE ;)

File: src/main/kotlin/com/sourcegraph/cody/ui/AutoGrowingTextArea.kt (hunk 4 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/ui/AutoGrowingTextArea.kt]: Just an IntelliJ suggestion accepted to simplify the code (inline assuagement and variable which could be private).

File: src/main/kotlin/com/sourcegraph/cody/statusbar/ReportCodyBugAction.kt (hunk 5 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/statusbar/ReportCodyBugAction.kt]: In the `CodyStatusBarActionGroup`.
What label param?

File: src/main/kotlin/com/sourcegraph/cody/initialization/TrialEndedNotification.kt (hunk 6 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/initialization/TrialEndedNotification.kt]: Turns out for `NotificationAction` it is not really needed, I will revert this.
In fact it looks it is needed only on classes which overrides `update` method, so I will do a few more reverts.

Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/cody/initialization/TrialEndedNotification.kt]: As for CI it's possible to run IJ inspection but I do not see an easy way to verify this in CI.

File: src/main/kotlin/com/sourcegraph/common/ui/DumbAwareBGTAction.kt (hunk 1 above)
Comment (pkukielka) [reply src/main/kotlin/com/sourcegraph/common/ui/DumbAwareBGTAction.kt]: Comments added.

**Review by mkondratek — APPROVED** [review]

**Review by mkondratek — APPROVED** [review]
_lepiej nie będzie_

**Pull request merged**